### DISCUSSION TITLE
How to implement Graphql  Connection

### DISCUSSION THREAD

**Discussion by hyzx86** [OrchardCMS/OrchardCore#8359] [Q&A]
Hi ，There is a problem here. How can I get a ConnectionType from this instance object:typeType ?

https://github.com/hyzx86/OrchardCore/blob/3ada9539cd5a5774fcc686e4b3d2785de25f2f0e/src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Queries/ContentTypeQuery.cs#L67

I see a similar problem here, but I don't understand how to operate it.
graphql-dotnet/graphql-dotnet#393

**Comment by hyzx86** [comment]
Associated with this issue #6939

**Comment by hyzx86** [comment]
now I have changed the code like bellow:  

```
                var typeType = new ContentItemType(_contentOptionsAccessor)
                {
                    Name = typeDefinition.Name,
                    Description = S["Represents a {0}.", typeDefinition.DisplayName]
                }; 
                var query = new ContentItemsFieldType(typeDefinition.Name, schema, _contentOptionsAccessor, _settingsAccessor)
                {
                    Name = typeDefinition.Name,
                    Description = S["Represents a {0}.", typeDefinition.DisplayName],
                    //UNDONE: how to get new ConnectionType from typeType?

                    ResolvedType = new ConnectionType<ContentItemType>()
                    {
                        Name = typeDefinition.Name,
                        Description = S["Represents a {0}.", typeDefinition.DisplayName]                        
                    }
                };
```
in this file: https://github.com/hyzx86/OrchardCore/blob/3ada9539cd5a5774fcc686e4b3d2785de25f2f0e/src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Queries/ContentTypeQuery.cs


change `ContentItemsFieldType.cs`:

https://github.com/hyzx86/OrchardCore/blob/3ada9539cd5a5774fcc686e4b3d2785de25f2f0e/src/OrchardCore/OrchardCore.ContentManagement.GraphQL/Queries/ContentItemsFieldType.cs#L47
 
Edit  ContentItemsFieldType.cs：

 `Type = typeof(ListGraphType<ContentItemType>);`  
to  
`Type = typeof(ConnectionType<ContentItemType>);`

and 
` Resolver = new LockedAsyncFieldResolver<Connection<ContentItem>>(Resolve);`

edit resolve method
```
            var totalCount = await contentItemsQuery.CountAsync();
            contentItemsQuery = PageQuery(contentItemsQuery, context, graphContext);

            var contentItems = await contentItemsQuery.ListAsync();

            foreach (var filter in filters)
            {
                contentItems = await filter.PostQueryAsync(contentItems, context);
            }
            //var endCursor = contentItems?.Last()?.ContentItemId;
            //var cursor = contentItems.Count() > 0 ? contentItems.Last()?.ContentItemId : null;
            return new Connection<ContentItem>()
            {
                TotalCount = totalCount,
                Edges = contentItems.Select(x => new Edge<ContentItem> { Cursor = x.ContentItemId, Node = x }).ToList(),
                PageInfo = new PageInfo
                {
                    HasNextPage = totalCount > first + skip,
                    HasPreviousPage = first > 0
                    //UNDONE:what about endCursor and cursor use for ? 
		 			//var endCursor = contentItems?.Last()?.ContentItemId; 
		            //var cursor = contentItems.Count() > 0 ? contentItems.Last()?.ContentItemId : null; 
                }
            };
```



It worked, but there were still some problems. that is schema cannot query for custom properties

**Comment by hyzx86** [comment]
![image](https://user-images.githubusercontent.com/15613121/105068026-608a6980-5abb-11eb-9c89-16df39e01ff2.png)

**Comment by EricBui619** [comment]
@hyzx86 
The solution right here: https://github.com/graphql-dotnet/graphql-dotnet/issues/393
Just clone the entire ConnectType and EdgeType from GraphQL .Net library and modify constructors to support instance-based resolvedType. I did it and it work well. I think it's the only solution right now
